### PR TITLE
Huggingface tokenizers

### PR DESCRIPTION
**Pull request by BenjaminBossan**
## Description

[Huggingface tokenizers](https://huggingface.co/docs/tokenizers/python/latest/) (especiall pre-trained ones) are widely used in
NLP tasks. However, it's not straightforward to integrate them with
skorch, there are a lot of small things to consider.

With this PR, we add `HuggingfaceTokenizer` and
`HuggingfacePretrainedTokenizer` classes, which take care of all these
small things. They provide a (mostly) sklearn compatible interface that
allows them to be used in a Pipeline with a skorch net (pending
acceptance of #840).

In this implementation, I did not opt for adding all existing options
available in Huggingface, instead focusing on a (still comprehensive)
list of (what seemed to me) most useful options.

## Implementation

How to handle the Huggingface dependencies (tokenizers and transformers)
could be up for debate. I think we should not strictly depend on them,
which is why they're not imported at root level, even though that
results in some ugly code.

Furthermore, I put the code inside a `hf.py` module. Maybe there are
better suggestions? If not, do we want to move the `AccelerateMixin` to
this module as well?

There is a minor issue at the moment with the tokenizer `Trainer` not
being pickleable. I mitigated this mostly by popping it from the state
and replacing it by None. This should be mostly acceptable because users
will tend to pickle trained transformers, and after training, the
trainer is no longer required.

However, this mitigation is still not enough when using sklearn's `clone`,
which recursively traverses all attributes to copy them. Possible
mitigations may be possible by overriding `__deepcopy__` or `__reduce_ex__`
methods, but those would probably still not allow cloning and then
fitting (e.g. for grid search).

I opened an issue, so that this problem will hopefully be addressed
soon:
https://github.com/huggingface/tokenizers/issues/941

## Documentation

I have a problem that `hf.py` does not appear in the TOC of the docs, at
least when I build them locally. Does anyone have an idea?

Also, if we want to proceed, do we want to a section on Huggingface integration
to the docs?

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
@ottonemo @thomasjpfan 

I made a rather big change to the implementation of `HuggingfaceTokenizer`.
Copied below is the corresponding commit message:

First of all, similar to the `NeuralNet` class, it is now possible to pass
uninitialized classes and have them initialized before fit. Most of the
machinery is a copy (with adjustments) of the same code form `NeuralNet`.

With these changes, it is now possible to perform grid searches etc. on
the individual parameters (instead of having to pass many instances).

Furthermore, by default the trainer is now being inferred. Thus it no
longer needs to be passed by the user and stored as an attribute on the
estimator. This mostly circumvents the pickling issues as well.

**Comment by BenjaminBossan** [comment]
Note (to myself): I want to add support for using a pretrained tokenizer to fit on new data using [this method](https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.PreTrainedTokenizerFast.train_new_from_iterator).

**Comment by BenjaminBossan** [comment]
> Note (to myself): I want to add support for using a pretrained tokenizer to fit on new data using [this method](https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.PreTrainedTokenizerFast.train_new_from_iterator).

Done.

Note on the state of being able to pickle the trainer: There was a [release](https://github.com/huggingface/tokenizers/releases) of version 0.12 of the tokenizers library but the release was yanked. Therefore, I'd argue we wait for a proper release before considering this as solved, and then pin the tokenizers version to that version or greater.

**Comment by BenjaminBossan** [comment]
@ottonemo I added a `tokenize` method: https://github.com/skorch-dev/skorch/pull/841/commits/74d8349ca2dde7feab99e91e7275e0a4f3b75c80

Could you please check if that corresponds to what we discussed?

For now, it is a bit convoluted because it first encodes, then decodes. Probably, we could use the individual components of the HF tokenizer to achieve the same goal directly, but it would be more work and it would probably create a dependency on implementation details of the tokenizer, so I chose the easy way for now.

**Comment by BenjaminBossan** [comment]
The CI error seems to be unrelated to the pushed change, probably it's due to the new sklearn release. I'll investigate.

**Comment by BenjaminBossan** [comment]
@ottonemo @thomasjpfan This PR has been lingering for some time. Is there anything I can do so that we can make progress?

File: skorch/hf.py [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+
+    Implements the ``vocabulary_`` attribute and the methods
+    ``get_feature_names``, ``transform``, and ``inverse_transform``.
+
+    Subclasses should implement the ``fit``  method.
+
+    """
+    @property
+    def vocabulary_(self):
+        check_is_fitted(self, ['fast_tokenizer_'])
```
Comment (adrinjalali) [inline skorch/hf.py:33]: Interesting. I would expect a property to raise `AttrubuteError` rather than `NotFittedError` if the estimator hasn't been fit. For instance, `logreg.coef_` raises an `AttrubuteError` when you haven't fit the model.

You could either leave it out, or use `@available_if` if you want (although I'm not sure if the latter works with a property).

File: skorch/hf.py [hunk 2]
```diff
@@ -0,0 +1,800 @@
...
+        Xt : transformers.tokenization_utils_base.BatchEncoding
+          A huggingface ``BatchEncoding`` instance. This is basically a
+          dictionary containing the ids of the tokens and some additional fields
+          (depending on the parameters), e.g. the attention mask. For this
+          reason, the output is not well suited to be used with normal sklearn
+          models, but it works with huggingface transformers and with skorch
+          nets.
+
+        """
+        check_is_fitted(self, ['fast_tokenizer_'])
```
Comment (adrinjalali) [inline skorch/hf.py:71]: you can also implement `__sklearn_is_fitted__` and call `check_is_fitted(self)` to do the same thing, and allow other estimators calling `check_is_fitted` on this estimator to do the same thing.

**Review by adrinjalali — COMMENTED** [review]
Please ignore the ones that make no sense, I'm not too familiar with skorch's internals.

**Review by ottonemo — APPROVED** [review]
I have tested this feature and to me it seems quite usable and intuitive. Code-wise there is nothing to add from my side (apart from the things we discussed already and you implemented).

File: skorch/hf.py (hunk 1 above)
Comment (BenjaminBossan) [reply skorch/hf.py:33]: Yes, you're right, `AttributeError` is more in line with expectations and with other sklearn transformers. I'll change the check accordingly.

File: skorch/hf.py [hunk 3]
```diff
@@ -0,0 +1,800 @@
...
+
+    Subclasses should implement the ``fit``  method.
+
+    """
+    @property
+    def vocabulary_(self):
+        check_is_fitted(self, ['fast_tokenizer_'])
+        return self.fast_tokenizer_.vocab
+
+    def get_feature_names(self):
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply skorch/hf.py:36]: Thanks for pointing to `get_feature_names_out`, I'll implement that method instead.

File: skorch/hf.py (hunk 2 above)
Comment (BenjaminBossan) [reply skorch/hf.py:71]: [Here](https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.check_is_fitted.html), it says

> If an estimator does not set any attributes with a trailing underscore, it can define a __sklearn_is_fitted__ method returning a boolean to specify if the estimator is fitted or not.

But from your comment I take that another advantage of implementing that method is that it allows other sklearn code to check if the estimator is fitted without the caller having to know what is needed for an estimator to be fitted, right?

File: skorch/hf.py [hunk 4]
```diff
@@ -0,0 +1,800 @@
...
+          (depending on the parameters), e.g. the attention mask. For this
+          reason, the output is not well suited to be used with normal sklearn
+          models, but it works with huggingface transformers and with skorch
+          nets.
+
+        """
+        check_is_fitted(self, ['fast_tokenizer_'])
+
+        # from sklearn, triggers a parameter validation
+        if isinstance(X, str):
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply skorch/hf.py:74]: I copied the code as e.g. here:

https://github.com/scikit-learn/scikit-learn/blob/baf0ea25d6dd034403370fea552b21a6776bef18/sklearn/feature_extraction/text.py#L1380-L1383

Do you think this is sufficient?

File: skorch/hf.py [hunk 5]
```diff
@@ -0,0 +1,800 @@
...
+
+    fast_tokenizer_ : transformers.PreTrainedTokenizerFast
+      If you want to extract the Huggingface tokenizer to use it without skorch,
+      use this attribute.
+
+    .. _tokenizers: https://huggingface.co/docs/tokenizers/python/latest/index.html
+
+    """
+    import transformers as _transformers
+    import tokenizers as _tokenizers
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply skorch/hf.py:295]: Although it's a bit ugly as is, I don't think it's enough to bring in another dependency.

File: skorch/hf.py [hunk 6]
```diff
@@ -0,0 +1,800 @@
...
+        self.max_length = max_length
+        self.return_tensors = return_tensors
+        self.return_attention_mask = return_attention_mask
+        self.return_token_type_ids = return_token_type_ids
+        self.return_length = return_length
+        self.pad_token = pad_token
+        self.verbose = verbose
+
+        self._kwargs = kwargs
+        vars(self).update(kwargs)
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply skorch/hf.py:333]: Yes, it's not something you would normally see in sklearn. However, I don't know if there is a better way to make it possible to have arbitrary parameters, e.g. `pre_tokenizer__delimiter`? I guess we could hide it inside `_kwargs` but having it only there can create its own problems, e.g. if a user tries to access `my_tokenizer.pre_tokenizer__delimiter` (instead of usng `get_params`) or if they set `my_tokenizer.pre_tokenizer__delimiter = 'foo'` directly (instead of using `set_params`). Yes, this is all fixable with some Python magic but I think having the attributes exist on the object directly is the more transparent and less error prone way.

File: skorch/hf.py [hunk 7]
```diff
@@ -0,0 +1,800 @@
...
+            suggestion = prefix + '__' + suffix
+            msgs.append(tmpl.format(key, suggestion))
+
+        if msgs:
+            full_msg = '\n'.join(msgs)
+            raise TypeError(full_msg)
+
+        return kwargs
+
+    def initialized_instance(self, instance_or_cls, kwargs):
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply skorch/hf.py:392]: Yes, this could be argued, but we have it on `NeuralNet` as well for convenience, so I think we should be consistent here.

File: skorch/hf.py [hunk 8]
```diff
@@ -0,0 +1,800 @@
...
+        self.tokenizer_.add_special_tokens([self.pad_token])
+        self.fast_tokenizer_ = self._transformers.PreTrainedTokenizerFast(
+            tokenizer_object=self.tokenizer_,
+            pad_token=self.pad_token,
+        )
+        self.fixed_vocabulary_ = False
+        return self
+
+    def __getstate__(self):
+        # The 'trainer' attribute cannot be pickled. To still allow pickling, we
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply skorch/hf.py:557]: I don't have any experience with `__reduce__`. The Python docs say:

> Although powerful, implementing `__reduce__()` directly in your classes is error prone. For this reason, class designers should use the high-level interface (i.e., `__getnewargs_ex__()`, `__getstate__()` and `__setstate__()`) whenever possible.

So maybe it's best to keep away from it?

Anyway, `tokenizers` has added pickle support, so we're hopefully just one release away from this problem vanishing (unfortunately, the last release was yanked).

File: skorch/hf.py (hunk 2 above)
Comment (BenjaminBossan) [reply skorch/hf.py:71]: Got it, I'll implement it this way.

File: skorch/hf.py [hunk 9]
```diff
@@ -0,0 +1,758 @@
...
+
+        """
+        # from sklearn, triggers a parameter validation
+        if isinstance(X, str):
+            raise ValueError(
+                "Iterable over raw text documents expected, string object received."
+            )
+        X = list(X)  # transformers tokenizer does not accept arrays
+
+        verbose = bool(getattr(self, 'verbose'))
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply skorch/hf.py:80]: Very good questions ^^
I changed it to use `self.verbose` directly.

**Comment by BenjaminBossan** [comment]
@adrinjalali Thanks for taking a look, it was really helpful. I implemented some of your suggestions and had some comments on the others, please have a look.

@ottonemo Thanks for the review. I made the change you suggested.

File: skorch/hf.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply skorch/hf.py:349]: Are these parameters of sub-estimators? `pre_tokenizer__delimiter` seems like one to me. In that case, you could maybe use `_BaseComposition` maybe? https://github.com/scikit-learn/scikit-learn/blob/6e7922e48a04f471b11e874fac67657941970794/sklearn/utils/metaestimators.py#L23 (and the usage in `Pipeline`)

File: skorch/hf.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply skorch/hf.py:311]: Note that the other packages have only copied the code in their respective `__init__.py` files instead of actually depending on this package ;)

File: skorch/hf.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply skorch/hf.py:90]: Yeah I guess for text there isn't much we can do.

File: skorch/hf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply skorch/hf.py]: fair enough.

**Review by adrinjalali — APPROVED** [review]
This LGTM, the rest is up to you :)

File: skorch/hf.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply skorch/hf.py:349]: The tokenizer is a sub-component, though not a `BaseEstimator`. In skorch, we stick close to `_BaseComposition._set_params` but have a few modifications to make certain things work, so we cannot just use `_BaseComposition`.
